Feature: Add shell context menu for widgets

hecksmosis opened the pull request:
<!-- 
🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨🚨
I ACKNOWLEDGE THE FOLLOWING BEFORE PROCEEDING:
1. PR may be deleted if it is not following the template
2. Try not to make duplicates. Do a quick search before posting
3. Add a clear title starting with "Feature:" or "Fix:"
-->

**Resolved / Related Issues**
Items resolved / related issues by this PR.
- Adds context menu items for widgets
- Closes #5888

**Validation**
How did you test these changes?
- [x] Built and ran the app
- [x] Tested the changes for accessibility

yaira2 commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -116,6 +122,13 @@ public QuickAccessWidget()
+			OpenInNewPaneCommand = new RelayCommand<FolderCardItem>(OpenInNewPane);
+			OpenPropertiesCommand = new RelayCommand<FolderCardItem>(OpenProperties);
+			PinToFavoritesCommand = new RelayCommand<FolderCardItem>(PinToFavorites);
+			UnpinFromFavoritesCommand = new RelayCommand<FolderCardItem>(UnpinFromFavorites);
```
I notice these commands are used in the other widgets as well, can we try to reuse them?

yaira2 commented:
Should we switch to CommandBarFlyout and display primary buttons for actions like "Open in new tab"?

ferrariofilippo commented on src/Files.App/UserControls/Widgets/RecentFilesWidget.xaml.cs:
```diff
@@ -89,6 +106,114 @@ public RecentFilesWidget()
+
+		private List<ContextMenuFlyoutItemViewModel> GetRecentItemMenuItems(RecentItem item)
+		{
+			return new List<ContextMenuFlyoutItemViewModel>()
+			{
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "RecentItemRemove/Text".GetLocalizedResource(),
+					Glyph = "\uF117",
+					GlyphFontFamilyName = "CustomGlyph",
+					Command = RemoveRecentItemCommand,
+					CommandParameter = item,
+					ShowItem = true
+				},
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "RecentItemClearAll/Text".GetLocalizedResource(),
+					Glyph = "\uF113",
+					GlyphFontFamilyName = "CustomGlyph",
+					Command = ClearAllItemsCommand,
+					ShowItem = true
+				},
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "RecentItemOpenFileLocation/Text".GetLocalizedResource(),
+					Glyph = "\uE737",
+					Command = OpenFileLocationCommand,
+					CommandParameter = item,
+					ShowItem = true
+				},
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "ShowMoreOptions".GetLocalizedResource(),
+					Glyph = "\xE712",
+					Items = new List<ContextMenuFlyoutItemViewModel>(),
+					ID = "ItemOverflow",
+					Tag = "ItemOverflow",
+					IsHidden = true,
+				}
+			}.Where(x => x.ShowItem).ToList();
```
You can remove `Where` since here `ShowItem` is a constant. 
Actually, you can remove `ShowItem = true` since that is its default value.

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -248,17 +414,12 @@ private void NotifyPropertyChanged([CallerMemberName] string propertyName = "")
-			var item = ((MenuFlyoutItem)sender).DataContext as FolderCardItem;
-			await NavigationHelpers.OpenPathInNewTab(item.Path);
-		}
+		private async void OpenInNewTab(FolderCardItem item)
+			=> await NavigationHelpers.OpenPathInNewTab(item.Path);
```
As above

yaira2 replied:
This is fine

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -248,17 +414,12 @@ private void NotifyPropertyChanged([CallerMemberName] string propertyName = "")
-			var item = ((MenuFlyoutItem)sender).DataContext as FolderCardItem;
-			CardNewPaneInvoked?.Invoke(this, new QuickAccessCardInvokedEventArgs { Path = item.Path });
-		}
+		private void OpenInNewPane(FolderCardItem item)
+			=> CardNewPaneInvoked?.Invoke(this, new QuickAccessCardInvokedEventArgs { Path = item.Path });
+		
```
As above

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -169,6 +189,152 @@ public bool ShowMultiPaneControls
+					Glyph = "\uE946",
+					Command = OpenPropertiesCommand,
+					CommandParameter = item,
+					ShowItem = true
```
You can remove this, `ShowItem = true` by default

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -169,6 +189,152 @@ public bool ShowMultiPaneControls
+					Glyph = "\uE737",
+					Command = OpenInNewWindowCommand,
+					CommandParameter = item,
+					ShowItem = true
```
You can remove this, `ShowItem = true` by default

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -169,6 +189,152 @@ public bool ShowMultiPaneControls
+					GlyphFontFamilyName = "CustomGlyph",
+					Command = OpenInNewTabCommand,
+					CommandParameter = item,
+					ShowItem = true
+				},
```
You can remove this, `ShowItem = true` by default

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -169,6 +189,152 @@ public bool ShowMultiPaneControls
+
+		private List<ContextMenuFlyoutItemViewModel> GetLocationItemMenuItems(FolderCardItem item)
+		{
+			var options = item.Item.MenuOptions;
+			var isPinned = item.Item.IsPinned;
```
`options` is nevere read

ferrariofilippo commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml:
```diff
@@ -7,133 +7,75 @@
+					Orientation="Horizontal" />
+			</ItemsRepeater.Layout>
+			<ItemsRepeater.ItemTemplate>
+				<DataTemplate x:DataType="local:FolderCardItem">
+                    <Button
+						Padding="8"
+						DataContext="{x:Bind}"
+						HorizontalAlignment="Stretch"
+						VerticalAlignment="Stretch"
+						HorizontalContentAlignment="Stretch"
+						VerticalContentAlignment="Stretch"
+						AutomationProperties.Name="{x:Bind AutomationProperties}"
+						Command="{x:Bind SelectCommand}"
+						CommandParameter="{x:Bind}"
+						CornerRadius="{StaticResource ControlCornerRadius}"
+                        RightTapped="Button_RightTapped"
+						PointerPressed="Button_PointerPressed"
+						Tag="{x:Bind Path}">
+                        <Grid
+							HorizontalAlignment="Stretch"
+							VerticalAlignment="Stretch"
+							RowSpacing="0">
+							<Grid.RowDefinitions>
```
Spacing seems off

ferrariofilippo commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml.cs:
```diff
@@ -156,61 +336,44 @@ private void NotifyPropertyChanged([CallerMemberName] string propertyName = "")
-			};
-			flyoutParent.Closed += flyoutClosed;
-		}
+		private async void OpenProperties(DriveCardItem item)
+			=> await FilePropertiesHelpers.OpenPropertiesWindowAsync(item.Item, associatedInstance);
```
Add back brackets

ferrariofilippo commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml.cs:
```diff
@@ -122,13 +138,177 @@ public DrivesWidget()
+					CommandParameter = item,
+					ShowItem = ShowMultiPaneControls
+				},
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "SideBarOpenInNewTab/Text".GetLocalizedResource(),
+					Glyph = "\uF113",
+					GlyphFontFamilyName = "CustomGlyph",
+					Command = OpenInNewTabCommand,
+					CommandParameter = item,
+					ShowItem = true
+				},
+				new ContextMenuFlyoutItemViewModel()
+				{
+					Text = "SideBarOpenInNewWindow/Text".GetLocalizedResource(),
+					Glyph = "\uE737",
+					Command = OpenInNewWindowCommand,
+					CommandParameter = item,
+					ShowItem = true
+				},
```
Here you can remove `ShowItem = true`

ferrariofilippo commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml.cs:
```diff
@@ -122,13 +138,177 @@ public DrivesWidget()
+					Text = "BaseLayoutContextFlyoutPropertiesFolder/Text".GetLocalizedResource(),
+					Glyph = "\uE946",
+					Command = OpenPropertiesCommand,
+					CommandParameter = item,
+					ShowItem = true
```
As above

ferrariofilippo commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml:
```diff
@@ -7,142 +7,62 @@
-					</Button.ContextFlyout>
-					<Grid
-						Margin="8"
+    <Grid>
+        <ItemsRepeater
+			x:Name="CardsList"
+			HorizontalAlignment="Stretch"
+			VerticalAlignment="Top"
+			ItemsSource="{x:Bind local:DrivesWidget.ItemsAdded, Mode=OneWay}">
+            <ItemsRepeater.Layout>
+                <UniformGridLayout
+					ItemsStretch="None"
+					MinColumnSpacing="8"
+					MinItemHeight="72"
+					MinItemWidth="240"
+					MinRowSpacing="8"
+					Orientation="Horizontal" />
+            </ItemsRepeater.Layout>
+            <ItemsRepeater.ItemTemplate>
+                <DataTemplate x:DataType="local:DriveCardItem">
+                    <Button
```
Spacing seems off in the entire file

ferrariofilippo reviewed (changes requested):
Also, `Button_RightTapped` and `LoadShellMenuItems` share almost all the code in all widgets. You may create an helper class to reduce code duplication

hecksmosis commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -309,12 +452,9 @@ private async void PinToFavorites_Click(object sender, RoutedEventArgs e)
-			_ = QuickAccessService.UnpinFromSidebar(item.Path);
-		}
-
+		private void UnpinFromFavorites(FolderCardItem item)
+			=> _ = QuickAccessService.UnpinFromSidebar(item.Path);
+		
```
Oh ok, didn't know about that

yaira2 commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml:
```diff
@@ -1,210 +1,131 @@
-							x:Phase="1"
-							Source="{x:Bind Thumbnail, Mode=OneWay}"
-							Stretch="Uniform" />
+    x:Class="Files.App.UserControls.Widgets.DrivesWidget"
```
This is wrong, spacing should be set to tabs.

hecksmosis replied:
I used XAML styler, how should I configure it?

hecksmosis commented:
I don't know if we can add the helper class for the Button_RightTapped because some use Grid

hecksmosis commented on src/Files.App/UserControls/Widgets/QuickAccessWidget.xaml.cs:
```diff
@@ -170,6 +175,148 @@ public bool ShowMultiPaneControls
+			e.Handled = true;
+		}
+
+		private async void LoadShellMenuItems(LocationItem item, CommandBarFlyout itemContextMenuFlyout, ContextMenuOptions options)
```
On it

hecksmosis replied:
Should be done

ferrariofilippo commented on src/Files.App/Helpers/ShellContextMenuHelper.cs:
```diff
@@ -210,5 +222,64 @@ void InstallFont(string path, bool asAdmin)
+						overflowItems.ForEach(i => flyoutItems.Add(i));
+					overflowItem.Visibility = overflowItems.Any() ? Visibility.Visible : Visibility.Collapsed;
+				}
+			}
+			
+			catch { }
```
Can you remove empty line?

ferrariofilippo commented on src/Files.App/Helpers/ShellContextMenuHelper.cs:
```diff
@@ -210,5 +222,64 @@ void InstallFont(string path, bool asAdmin)
+				}
+				
+				var shiftPressed = InputKeyboardSource.GetKeyStateForCurrentThread(VirtualKey.Shift).HasFlag(CoreVirtualKeyStates.Down);
+				var shellMenuItems = await ContextFlyoutItemHelper.GetItemContextShellCommandsAsync(workingDir: null,
+					new List<ListedItem>() { new ListedItem(null) { ItemPath = path } }, shiftPressed: shiftPressed, showOpenMenu: false, default);
```
var shellMenuItems = await ContextFlyoutItemHelper.GetItemContextShellCommandAsync(
         workingDir: null,
         new List<ListedItem>() { new ListedItem(null) { ItemPath = path } },
         shiftPressed: shiftPressed,
         showOpenMenu: false,
         default);

ferrariofilippo reviewed (dismissed):
LGTM

yaira2 commented:
Just took a quick look and this much closer to what I had in mind. Will do a full review later but it's looking good.

yaira2 commented:
@hecksmosis functionality looks good too, I wonder if we should move options like properties and open with to the main menu. Just trying to think of ways to make it a little more consistent with the right click menu in the different layout modes.

hecksmosis commented:
> @hecksmosis functionality looks good too, I wonder if we should move options like properties and open with to the main menu. Just trying to think of ways to make it a little more consistent with the right click menu in the different layout modes.

I think properties is in the main menu already

hecksmosis commented:
Do you want to add properties to recent files?

yaira2 commented:
Yes, properties and open with.

yaira2 commented:
![image](https://user-images.githubusercontent.com/39923744/217860456-50ccf2aa-133b-4c75-ae80-eb3e27543b6a.png)

Any idea why it's showing these icons?

hecksmosis commented:
> Yes, properties, open with and add to favorites.

Add to favorites wouldn't make sense as only files appear there @yaira2

yaira2 commented:
Good point, comes back to the fact that we don't show pinned files yet. We can probably hide the menu item for that.

hecksmosis commented:
The icons are because I didn't notice

hecksmosis commented:
I fixed the icons and added the open with menu, but I haven't been able to add the properties menu as it shows incomplete, help would be appreciated in that aspect.

yaira2 commented:
I think it's okay for now, I'll try to finish reviewing today.

d2dyno1 commented on src/Files.App/ServicesImplementation/QuickAccessService.cs:
```diff
@@ -52,5 +53,10 @@ public async Task UnpinFromSidebar(string[] folderPaths)
 			App.QuickAccessManager.UpdateQuickAccessWidget?.Invoke(this, new ModifyQuickAccessEventArgs(folderPaths, false));
 		}
+
+		public bool IsItemPinned(string folderPath)
```
Maybe `ILocatableFolder`? 🤔

d2dyno1 commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml.cs:
```diff
@@ -156,60 +248,16 @@ private void NotifyPropertyChanged([CallerMemberName] string propertyName = "")
 		}
 
-		private async void EjectDevice_Click(object sender, RoutedEventArgs e)
+		private async void EjectDevice(DriveCardItem item)
```
`async Task`

d2dyno1 commented on src/Files.App/UserControls/Widgets/DrivesWidget.xaml.cs:
```diff
@@ -253,14 +301,13 @@ public bool ShowMultiPaneControls
 		}
 
-		private async void OpenInNewPane_Click(object sender, RoutedEventArgs e)
+		private async void OpenInNewPane(DriveCardItem item)
```
Same here

d2dyno1 commented on src/Files.App/UserControls/Widgets/FileTagsWidget.xaml.cs:
```diff
@@ -51,6 +62,49 @@ private async void FileTagItem_ItemClick(object sender, ItemClickEventArgs e)
 				await itemViewModel.ClickCommand.ExecuteAsync(null);
 		}
 
+		private void Item_RightTapped(object sender, RightTappedRoutedEventArgs e)
```
Idk but it feels to me as code duplication. Can that be avoided?

hecksmosis replied:
I haven't found a way, the XAML elements are different

d2dyno1 commented on src/Files.Backend/Services/IQuickAccessService.cs:
```diff
@@ -39,5 +39,7 @@ public interface IQuickAccessService
 		/// <returns></returns>
 		Task UnpinFromSidebar(string[] folderPaths);
+
+		bool IsItemPinned(string folderPath);
```
Could you add docs :)

d2dyno1 reviewed (commented):
Some nits 🙂

hecksmosis commented:
@d2dyno1 Is it ok now?

d2dyno1 commented:
/azp run

yaira2 commented:
It's not loading on recent files for me but I don't want to hold this back so let's merge as is.
